### PR TITLE
Optimization

### PR DESCRIPTION
**Pull request by Nicolasdejean**
little optimizations in TheThingsNetwork.cpp

### CODE REVIEW

File: src/TheThingsNetwork.cpp [hunk 1]
```diff
@@ -177,10 +177,7 @@ bool TheThingsNetwork::join(int8_t retries, uint32_t retryDelay) {
   str.concat(F("mac set deveui "));
   str.concat(devEui);
   sendCommand(str);
-  while (retries != 0) {
-    if (retries > 0) {
-      retries--;
-    }
+  for (size_t i = retries; i--; ) {
```
Comment (johanstokking) [inline src/TheThingsNetwork.cpp]: You can cut even `i`:

`while (--retries >= 0)`

File: src/TheThingsNetwork.h [hunk 2]
```diff
@@ -6,13 +6,130 @@
...
+const char sys_get[] PROGMEM = "get";
+const char sys_get_ver[] PROGMEM = "ver";
+const char sys_get_vdd[] PROGMEM = "vdd";
+const char sys_get_hweui[] PROGMEM = "hweui";
+const char sys_set_get_nvm[] PROGMEM = "nvm";
+const char sys_set_pindig[] PROGMEM = "pindig";
+
+const char* const sys_table[] PROGMEM = {sys_prefix,sys_sleep,sys_reset,sys_erase_fw,sys_factory_rst,sys_set,sys_get,sys_get_ver,sys_get_vdd,sys_get_hweui,sys_set_get_nvm,sys_set_pindig};
+
+#define SYS_PREFIX 0
```
Comment (FokkeZB) [inline src/TheThingsNetwork.h]: All these defines will be public. Can't we move this to the `cpp` file?

Comment (Nicolasdejean) [inline src/TheThingsNetwork.h]: Sure we can, but won't the `cpp` file be kind of messy after ?

Comment (FokkeZB) [inline src/TheThingsNetwork.h]: It will, but we should only make stuff public that the public needs.

**Comment by FokkeZB** [comment]
Just used it and works great, with 53% storage used (instead of 95%!)

**Comment by johanstokking** [comment]
@FokkeZB merge when you think it's ready

File: src/TheThingsNetwork.h [hunk 3]
```diff
@@ -6,13 +6,35 @@
...
 #define TTN_DEFAULT_FSB 2
-#define TTN_RETX 7
+#define TTN_RETX "7"
 
-#define TTN_PWRIDX_868 1
-#define TTN_PWRIDX_915 5
+#define TTN_PWRIDX_868 "1"
+#define TTN_PWRIDX_915 "5"
+
+#define TTN_BUFFER_SIZE 300
```
Comment (FokkeZB) [inline src/TheThingsNetwork.h]: Please check if there are more defines that can be moved to `.cpp` to not be public. Only those needed to change default settings should be here, and also added to `keywords.txt`.

**Pull request merged**